src/v/cluster/controller_api.cc

```diff
@@ -174,21 +175,39 @@ controller_api::get_reconciliation_state(model::ntp ntp) {
     const auto shards = boost::irange<ss::shard_id>(0, ss::smp::count);
     for (auto shard : shards) {
         auto local_deltas = co_await get_remote_core_deltas(ntp, shard);
-
-        std::transform(
-          local_deltas.begin(),
-          local_deltas.end(),
-          std::back_inserter(ops),
-          [shard](controller_backend::delta_metadata& m) {
-              return backend_operation{
-                .source_shard = shard,
-                .p_as = std::move(m.delta.new_assignment),
-                .type = m.delta.type,
-                .current_retry = m.retries,
-                .last_operation_result = m.last_error,
-                .revision_of_operation = model::revision_id(m.delta.offset),
-              };
-          });
+        for (auto& m : local_deltas) {
+            ss::visit(
+              m.delta.get_data(),
+              [&ops, shard, &m](delta_reconfiguration_data& data) {
+                  ops.push_back(backend_operation{
+                    .source_shard = shard,
+                    .p_as = std::move(data.target_assignment),
+                    .type = m.delta.type(),
+                    .current_retry = m.retries,
+                    .last_operation_result = m.last_error,
+                    .revision_of_operation = m.delta.revision(),
+                  });
+              },
+              [&ops, shard, &m](delta_finish_update_data& data) {
+                  ops.push_back(backend_operation{
+                    .source_shard = shard,
+                    .p_as = data.target_assignment,
+                    .type = m.delta.type(),
+                    .current_retry = m.retries,
+                    .last_operation_result = m.last_error,
+                    .revision_of_operation = m.delta.revision(),
+                  });
+              },
+              [&ops, shard, &m](auto&) {
+                  ops.push_back(backend_operation{
+                    .source_shard = shard,
+                    .type = m.delta.type(),
+                    .current_retry = m.retries,
+                    .last_operation_result = m.last_error,
+                    .revision_of_operation = m.delta.revision(),
+                  });
+              });
+        }
```

Comment:
nit: could be shortened to something like
```
auto assignment = ss::visit(...
ops.push_back(backend_operation{
...
});
```
because assignment is really the only field that we need access to variants for, everything else is the same.